File: src/php/OasParserFactory.php
<?php

namespace TuEmpresa\LaravelAiApiIntegration\Infrastructure\Services\OasParser;

use TuEmpresa\LaravelAiApiIntegration\Domain\Interfaces\OasParserInterface;
use TuEmpresa\LaravelAiApiIntegration\Domain\Exceptions\OasParserException;
use Symfony\Component\Yaml\Yaml;

class OasParserFactory {
    public static function create(string $filePath): OasParserInterface {
        $extension = pathinfo($filePath, PATHINFO_EXTENSION);
        
        if (!in_array($extension, ['json', 'yaml', 'yml'])) {
            throw new OasParserException("Unsupported file format: $extension");
        }

        try {
            $content = file_get_contents($filePath);
            if ($content === false) {
                throw new OasParserException("Unable to read file: $filePath");
            }

            $data = self::parseContent($content, $extension);
            return self::createParserByVersion($data);
        } catch (\Exception $e) {
            throw new OasParserException("Error creating OAS parser: " . $e->getMessage(), 0, $e);
        }
    }

    private static function parseContent(string $content, string $extension): array {
        if ($extension === 'json') {
            $data = json_decode($content, true, 512, JSON_THROW_ON_ERROR);
        } else {
            $data = Yaml::parse($content);
        }

        if (!is_array($data)) {
            throw new OasParserException("Invalid OpenAPI specification: file content is not a valid JSON/YAML.");
        }

        if (!isset($data['openapi']) || !is_string($data['openapi'])) {
            throw new OasParserException("Invalid OpenAPI specification: 'openapi' version not found or invalid.");
        }

        return $data;
    }

    private static function createParserByVersion(array $data): OasParserInterface {
        $version = $data['openapi'];
        if (strpos($version, '3.0') === 0) {
            return new Oas30Parser();
        } elseif (strpos($version, '3.1') === 0) {
            return new Oas31Parser();
        } else {
            throw new OasParserException("Unsupported OpenAPI version: $version");
        }
    }
}

File: src/infrastructure/services/oas_parser/oas_parser_factory.py
import yaml
import json
from pathlib import Path
from src.domain.interfaces.oas_parser_interface import OasParserInterface
from src.domain.exceptions.oas_parser_exception import OasParserException
from src.infrastructure.services.oas_parser.oas30_parser import Oas30Parser
from src.infrastructure.services.oas_parser.oas31_parser import Oas31Parser

class OasParserFactory:
    @staticmethod
    def create(file_path: str) -> OasParserInterface:
        path = Path(file_path)
        if not path.exists():
            raise OasParserException(f"File not found: {file_path}")

        try:
            content = path.read_text()
            data = OasParserFactory._parse_content(content, path.suffix)
            return OasParserFactory._create_parser_by_version(data)
        except Exception as e:
            raise OasParserException(f"Error creating OAS parser: {str(e)}")

    @staticmethod
    def _parse_content(content: str, extension: str) -> dict:
        if extension == '.json':
            data = json.loads(content)
        elif extension in ['.yaml', '.yml']:
            data = yaml.safe_load(content)
        else:
            raise OasParserException(f"Unsupported file format: {extension}")

        if not isinstance(data, dict):
            raise OasParserException("Invalid OpenAPI specification: file content is not a valid JSON/YAML.")

        if 'openapi' not in data or not isinstance(data['openapi'], str):
            raise OasParserException("Invalid OpenAPI specification: 'openapi' version not found or invalid.")

        return data

    @staticmethod
    def _create_parser_by_version(data: dict) -> OasParserInterface:
        version = data['openapi']
        if version.startswith('3.0'):
            return Oas30Parser()
        elif version.startswith('3.1'):
            return Oas31Parser()
        else:
            raise OasParserException(f"Unsupported OpenAPI version: {version}")
